### DISCUSSION TITLE
run.cpu_bound "Can't pickle local object" error

### DISCUSSION THREAD

**Discussion by johancj** [zauberzeug/nicegui#2221] [Q&A] [answered]
### Question

I have blocking code which I would like to run using `run.cpu_bound` inside a page (because I have variables located here which `heavy_computation` uses in real life).

When I run the code below, I get the following error:
`AttributeError: Can't pickle local object 'main_page.<locals>.heavy_computation'`

Is there any way of running blocking code which is located within a `page`?

Here is a script for reproducing the error based on https://github.com/zauberzeug/nicegui/blob/main/examples/progress/main.py, but `heavy_computation` is moved to within the page:

```python
#!/usr/bin/env python3
import time
from multiprocessing import Manager, Queue

from nicegui import run, ui

@ui.page('/')
def main_page():
    def heavy_computation(q: Queue) -> str:
        """Run some heavy computation that updates the progress bar through the queue."""
        n = 50
        for i in range(n):
            # Perform some heavy computation
            time.sleep(0.1)

            # Update the progress bar through the queue
            q.put_nowait(i / n)
        return 'Done!'
    async def start_computation():
        progressbar.visible = True
        result = await run.cpu_bound(heavy_computation, queue)
        ui.notify(result)
        progressbar.visible = False

    # Create a queue to communicate with the heavy computation process
    queue = Manager().Queue()
    # Update the progress bar on the main process
    ui.timer(0.1, callback=lambda: progressbar.set_value(queue.get() if not queue.empty() else progressbar.value))

    # Create the UI
    ui.button('compute', on_click=start_computation)
    progressbar = ui.linear_progress(value=0).props('instant-feedback')
    progressbar.visible = False

ui.run()
```

**Comment by rodja** [comment, marked as answer]
`run.cpu_bound` needs to execute the function in a separate process. For this it needs to transfer the whole state of the passed function to the process (which is done with pickle). We typically solve this in our own code by creating static methods (or free functions) which get all the data as simple parameters (eg. no class/ui logic).

**Reply by falkoschindler** [reply]
I find it interesting that the location of `heavy_computation` (within or outside of `main_page()`) makes such a difference for pickling. In our official example this function _can_ be pickled. But when moved inside the page function, it suddenly can't.

**Reply by rodja** [reply]
That is because the whole closure needs to be available in the subprocess. Python can not know which parts of the closure you are accessing.

**Reply by johancj** [reply]
Alright, thank you so much for extremely quick response as always! 💯

**Reply by zhangguolin333** [reply]
@johancj @falkoschindler   So how did it end up working out? I don't understand the above solution. Can I post the code after the solution? I have the same question that puzzles me, thank you ~

**Reply by rodja** [reply]
@zhangguolin333 if it really is the same problem but you could not fix it with the informations here provided it might be good to post it here. But please, try to create a minimum reproducible example. This would allow us and the community to help more efficiently. Thanks!
[Why is it important to provide a minimal reproducible example?](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-it-important-to-provide-a-minimal-reproducible-example)

**Reply by zhangguolin333** [reply]
@rodja  as in the above example, I seem to know that **`heavy_computation`** is defined as a free function that no longer depends on the local scope of the **`main_page`** function. In this way, **`run.cpu_bound`** can correctly serialize this function using **`pickle`** and execute it in separate processes
